# Session 3 — Project Tasks — P4A 2026
**Group 23**

In [16]:
import os
import pandas as pd
import numpy as np
import sqlalchemy as sa
from dotenv import load_dotenv

load_dotenv() 

DB_HOST = "3de0dac0-8513-4220-9ee7-414dc040c138.bn2a2uid0up8mv7mv2ig.databases.appdomain.cloud"
DB_PORT = "31131"
DB_NAME = "linkedin_jobs"

url = sa.engine.URL.create(
    drivername="mysql+pymysql",
    username=os.getenv("DB_USER"),
    password=os.getenv("DB_PASSWORD"),
    host=os.getenv("DB_HOST", DB_HOST),
    port=int(os.getenv("DB_PORT", DB_PORT)),
    database=os.getenv("DB_NAME", DB_NAME),
)

engine = sa.create_engine(url, connect_args={"ssl": {"check_hostname": False}})
print(url)  # password appears as ***

mysql+pymysql://a20254348:***@3de0dac0-8513-4220-9ee7-414dc040c138.bn2a2uid0up8mv7mv2ig.databases.appdomain.cloud:31131/linkedin_jobs


In [25]:
pd.read_sql("SHOW TABLES", engine)

,Tables_in_linkedin_jobs
0,companies_companies
1,companies_company_industries
2,companies_company_specialities
3,companies_employee_counts
4,jobs_benefits
5,jobs_job_industries
6,jobs_job_skills
7,jobs_salaries
8,mappings_industries
9,mappings_skills


## Task 1 — Load the staging table

In [29]:
pd.read_sql("SELECT COUNT(*) FROM jobs_benefits", engine)


,COUNT(*)
0,67943


In [50]:
df_postings = pd.read_sql("""
    SELECT p.*, GROUP_CONCAT(b.type SEPARATOR ';') AS benefits
    FROM postings p
    LEFT JOIN jobs_benefits b ON p.job_id = b.job_id
    GROUP BY p.job_id
""", engine)

print("Shape:", df_postings.shape)
print("Coluna 'benefits' presente:", "benefits" in df_postings.columns)
print("% com benefits:", df_postings['benefits'].notna().mean() * 100)



Shape: (40059, 32)
Coluna 'benefits' presente: True
% com benefits: 18.332958885643674


## Task 2 — Explore the benefits column

In [49]:
pct_with_benefits = df_postings['benefits'].notna().mean() * 100
print(f"{pct_with_benefits:.1f}% of postings have at least one benefit listed")
display(df_postings[['job_id', 'title', 'benefits']].head(100))





18.3% of postings have at least one benefit listed


,job_id,title,benefits
0,921716,Marketing Coordinator,NaN
1,1218575,Respiratory Therapist,NaN
2,1829192,Mental Health Therapist/Counselor,NaN
3,2264355,Worship Leader,NaN
4,9615617,Inside Customer Service Associate,NaN
...,...,...,...
95,2920450495,Service Coordinator,NaN
96,2934066464,Fundraising Associate,NaN
97,2954591060,Social Media Coordinator,NaN
98,2957460032,Professional Window Cleaning Technician,NaN


In [34]:
top10_benefits = (
    df_postings['benefits']
    .dropna()
    .str.split(';')
    .explode()
    .str.strip()
    .value_counts()
    .head(10)
)
print("Top 10 most common benefits:")
print(top10_benefits)

Top 10 most common benefits:
benefits
401(k)                  5912
Medical insurance       2657
Vision insurance        2420
Disability insurance    2027
Dental insurance        1873
Tuition assistance       635
Commuter benefits        584
Paid maternity leave     507
Paid paternity leave     437
Pension plan             212
Name: count, dtype: int64


**Insight:** Only 18.3% of job postings include at least one benefit listed, indicating that the majority of companies on LinkedIn do not explicitly advertise perks in their postings. This low coverage limits the representativeness of any benefit-based analysis, as conclusions drawn may only reflect a small and potentially biased subset of all job postings.

## Task 3 — Load the companies data

In [51]:
df_companies = pd.read_sql("SELECT * FROM companies_companies", engine)
print(f"Shape: {df_companies.shape}")
df_companies.head()

Shape: (24473, 10)


,company_id,name,description,company_size,state,country,city,zip_code,address,url
0,1009,IBM,"At IBM, we do more than work. We create. We cr...",7.0,NY,US,"Armonk, New York",10504,International Business Machines Corp.,https://www.linkedin.com/company/ibm
1,1016,GE HealthCare,Every day millions of people feel the impact o...,7.0,0,US,Chicago,0,-,https://www.linkedin.com/company/gehealthcare
2,1025,Hewlett Packard Enterprise,Official LinkedIn of Hewlett Packard Enterpris...,7.0,Texas,US,Houston,77389,1701 E Mossy Oaks Rd Spring,https://www.linkedin.com/company/hewlett-packa...
3,1028,Oracle,We’re a cloud technology company that provides...,7.0,Texas,US,Austin,78741,2300 Oracle Way,https://www.linkedin.com/company/oracle
4,1033,Accenture,Accenture is a leading global professional ser...,7.0,0,IE,Dublin 2,0,Grand Canal Harbour,https://www.linkedin.com/company/accenture


## Task 4 — Assess data quality

In [52]:
for col in ['company_size', 'country', 'city']:
    pct_null = df_companies[col].isna().mean() * 100
    print(f"{col}: {pct_null:.1f}% null")

company_size: 11.3% null
country: 0.0% null
city: 0.0% null


**Comment:** `country` and `city` present no missing values (0.0% null),making them fully reliable for geographic analysis. `company_size` has a null rate of 11.3%, which is relatively low the vast majority of companies have this field populated, so size-based segmentation remains valid, though conclusions should acknowledge that a small subset of companies is excluded.

## Task 5 — Filter to US companies

In [53]:
df_us = df_companies[df_companies['country'] == 'US']
print("Number of companies in the USA:", len(df_us))

Number of companies in the USA: 21635


## Task 6 — Count companies by size

In [58]:
# Verificar se df_us tem linhas
print("Shape df_us:", df_us.shape)

# Ver os valores únicos de company_size
print("Valores únicos:", df_us['company_size'].unique())

# Ver quantos não-nulos existem
print("Não-nulos:", df_us['company_size'].notna().sum())

# Ver as primeiras linhas
display(df_us.head())

# Ver todos os valores únicos de country no df original
print(df_companies['country'].value_counts().head(20))

Shape df_us: (21635, 10)
Valores únicos: [ 7.  5.  4.  6.  2.  1.  3. nan]
Não-nulos: 19329


,company_id,name,description,company_size,state,country,city,zip_code,address,url
0,1009,IBM,"At IBM, we do more than work. We create. We cr...",7.0,NY,US,"Armonk, New York",10504,International Business Machines Corp.,https://www.linkedin.com/company/ibm
1,1016,GE HealthCare,Every day millions of people feel the impact o...,7.0,0,US,Chicago,0,-,https://www.linkedin.com/company/gehealthcare
2,1025,Hewlett Packard Enterprise,Official LinkedIn of Hewlett Packard Enterpris...,7.0,Texas,US,Houston,77389,1701 E Mossy Oaks Rd Spring,https://www.linkedin.com/company/hewlett-packa...
3,1028,Oracle,We’re a cloud technology company that provides...,7.0,Texas,US,Austin,78741,2300 Oracle Way,https://www.linkedin.com/company/oracle
5,1035,Microsoft,Every company has a mission. What's ours? To e...,7.0,Washington,US,Redmond,98052,1 Microsoft Way,https://www.linkedin.com/company/microsoft


country
US    21635
0       727
GB      620
CA      258
IN      157
DE      128
FR      118
CH       93
AU       73
NL       64
SE       48
CN       47
IT       46
DK       42
ES       31
SG       30
BE       29
IE       28
JP       23
IL       21
Name: count, dtype: int64


In [44]:
size_counts = df_us['company_size'].value_counts().sort_index()
print("Companies per size (1=smallest, 7=largest):")
print(size_counts)
print(f"\nMost common size:  {size_counts.idxmax()}")
print(f"Least common size: {size_counts.idxmin()}")

Companies per size (1=smallest, 7=largest):
company_size
1.0    3866
2.0    4584
3.0    2892
4.0    2160
5.0    3467
6.0     903
7.0    1457
Name: count, dtype: int64

Most common size:  2.0
Least common size: 6.0


## Task 7 — Compute coverage for company size

In [45]:
coverage = df_us['company_size'].notna().mean() * 100
print(f"Company size coverage for US companies: {coverage:.1f}%")

Company size coverage for US companies: 89.3%


**Comment:** Since only a portion of US companies have a non-null `company_size`, the size distribution in Task 6 only reflects companies that chose to report their size conclusions may not generalise to all US companies on LinkedIn.

## Task 8 — Identify the top 10 countries

In [46]:
top_countries = df_companies['country'].value_counts().head(10)
print("Top 10 countries by number of companies:")
print(top_countries)

Top 10 countries by number of companies:
country
US    21635
0       727
GB      620
CA      258
IN      157
DE      128
FR      118
CH       93
AU       73
NL       64
Name: count, dtype: int64


**Comment:** The dataset is heavily concentrated in the United States (21635 companies), followed at a considerable distance by the United Kingdom (620) and Canada (258). Considering that the total sample contains 24473 records, this dominance is not surprising and highlights that the data mainly reflects LinkedIn’s strong user presence in English-speaking countries, particularly in North America, rather than the actual global job market distribution. Additionally, 727 records contain “0” as the country value, indicating a data quality issue that should be taken into account in any geographic analysis.